### ISSUE TITLE
LoanPipelineCursor should allow start+count > total

### ISSUE THREAD

**Issue by hfcjweinstock** [EncompassRest/EncompassRest#24]
if start + count is greater than the total remaining items, only return the remaining items?

**Comment by TylerBrinkley** [comment]
The total count is available in the `Count` property so users should know how many they can retrieve.

Afaik, the SDK does this bounds check too.

**Comment by hfcjweinstock** [comment]
the SDK has "Start" and "Limit" not a count

**Comment by TylerBrinkley** [comment]
The API has start and limit which are both exposed in the raw method. The SDK has startIndex and count.

So essentially, I initially tried to design the .NET API

**Comment by hfcjweinstock** [comment]
plus, the API call may not return the count # of loans if the back end determines that is too large.  I think the upper bounds check is not needed.

**Comment by hfcjweinstock** [comment]
another thought on this is that the result may not match count.  if the api returns less than the limit, that is ok.

**Comment by TylerBrinkley** [comment]
So essentially, when I first tried to design the .NET API for this feature I tried to make it so that it could retrieve results on the creation call as is supported in the API. So I created the API to look like this.

```c#
public sealed class Pipeline
{
    public LoanPipelineCursor CreateCursor(PipelineParameters parameters);
}
public sealed class LoanPipelineCursor
{
    public string CursorId { get; } // Initially null
    public int Count { get; } // Initially -1 until the first call
    public Task InitializeAsync(); // Needed if you want the Count or to create the Cursor on the server before getting any items
    public Task<LoanPipelineData> GetItemAsync(int start);
    public Task<List<LoanPipelineData>> GetItemsAsync(int start, int? limit);
}
```

The reasons I didn't like this was that `CreateCursor` doesn't actually create the cursor on the server as well as the fact that if you specify a `start` that is not less than `Count` then the Encompass API will throw an error like below.

```json
{
    "summary": "Bad Request",
    "details": "The 0-based 'start' position exceeds the total number of loans (558) that can be returned based on the given filter."
}
```

Also if there are no results then a CursorId is not returned.

I suppose we could add overloads for `GetItems` with a `int? limit` parameter instead of `count` but still retain the `count` version for those that need an exact number of results whether with one API request or multiple. What do you think?

**Comment by TylerBrinkley** [comment]
On second thought, maybe we should just remove the `count` version and make the user make multiple API requests if needed.

**Comment by hfcjweinstock** [comment]
my issue is on the getItems - the start and count parameters. 

start needs to be checked. i get that. i also understand that start falling outside the bounds of the total Count is an issue.

but the get items count is a "limit" in the API.  if you put 500 in there, it will do its best to return 500.  if you manage to pull less because of size limits or running out of loans then you will get back less than 500.  The count parameter just needs to be >0. on the usage side, you need to check how many loans you get back.

**Comment by hfcjweinstock** [comment]
i thought the create cursor does create the cursor on the server.

**Comment by TylerBrinkley** [comment]
It does now, I initially wanted to support retrieval of the first items on creation of the cursor on the server but that muddied up the API.

I agree I'll submit a PR to change `int count` to `int? limit` as is in the `GetItemsRawAsync` method. Do you think there's a need for `GetItemAsync`?

**Comment by hfcjweinstock** [comment]
and it should be updated to have Fields be a parameter.  you can change that on the fly.

**Comment by hfcjweinstock** [comment]
one other bit - the cursor is not Forward only.  it is pull anything you want.

**Comment by TylerBrinkley** [comment]
We may also want to support creating a `LoanPipelineCursor` object from an existing `CursorId`.

**Comment by hfcjweinstock** [comment]
Thought about that.  it is more complicated due to the "count"

**Comment by hfcjweinstock** [comment]
closed with #40
